src/rosdiscover/observer/ros1.py

```diff
@@ -0,0 +1,304 @@
+
+    def add_unfinished_to_context(self, nc: NodeContext):
+        """Add topics from incomplete action server/client to node context
+
+        Parameters
+        ----------
+        nc: NodeContext
+            The node to add the topics to
+        """
+        assert not self.is_complete()
+        if self.goal is not None:
+            self.add(nc, self.goal, True)
+        if self.cancel is not None:
+            self.add(nc, self.cancel, True)
+        if self.status is not None:
+            self.add(nc, self.status, False)
+        if self.feedback is not None:
+            self.add(nc, self.feedback, False)
+        if self.result is not None:
+            self.add(nc, self.result, False)
+
+
+def get_action_candidate(node: str,
+                         action_name: str,
+                         existing: Dict[str, Dict[str, ActionCandidate]],
+                         is_server: bool) -> ActionCandidate:
+    """Gets the action candidate if it exsits
+
+    Parameters
+    ----------
+    node: str
+        The name of the node that the action might belong to
+    action_name: str
+        The name of the action to look for or create
+    existing: Dict[str, Dict[str, ActionCandidate]]
+        The existing known actions. This is updated if a new candidate is created
+    is_server:
+        Whether this should be an action server
+
+    Returns
+    -------
+    ActionCandidate
+        The new or existing action candidate
+
+    """
+    if node in existing:
+        if action_name in existing[node]:
+            a_s = existing[node][action_name]
+        else:
+            a_s = ActionCandidate(action_name, is_server)
+            existing[node][action_name] = a_s
+    else:
+        a_s = ActionCandidate(action_name, is_server)
+        existing[node] = {action_name: a_s}
+    return a_s
+
+
+def action_candidate(node: str,
+                     topic: str,
+                     fmt: str,
+                     publishes: bool,
+                     existing: Dict[str, Dict[str, ActionCandidate]]):
+    """
+    Check if the topic might belong to an action. The parameter
+    :code:'existing' is updated with a new candidate if it is created. The topic is added to
+    the action if it is meant to be.
+
+    Parameters
+    ----------
+    node: str
+        The name of the node that the action might be associated with.
+    topic: str
+        The name of the topic that could be part of an action.
+    fmt: str
+        The topic format associated with the topic.
+    publishes: bool
+        Indicates whether the topic is published by the node (subscribed otherwise).
+    existing: Dict[str, Dict[str, ActionCandidate]]
+        The existing candidates. This will be updated if the topic might be part of
+        an action and the candidate isn't already known.
+    """
+    goal = _GOAL.match(topic)
+    format_match = _GOAL_FMT.match(fmt)
+
+    a_c = None
+    if goal is not None and format_match is not None:
```

Comment:
This method is pretty complex and the control flow is a little tricky to follow. It seems that there's a degree of duplicated code here. Is it possible to refactor this?